aips/aip-resource-access.md

```diff
@@ -0,0 +1,196 @@
+
+
+## Alternative solutions
+
+In Solana, a transaction must declare upfront the accounts it reads or writes from. Similarily, in Sui Move a transaction declares what objects it accesses. Using Aptos Objects as basis for what is proposed here has been discussed. Arguably, the approach described here is closer to the core Move language, as it does not depend on library extensions like objects. The approach is also more general, as it allows a variety of different access patterns, including negation.
+
+
+## Specification
+
+### Source Language
+
+The existing `acquires T` notation in Move is replaced by an extended syntax:
+
+```move
+fun f() acquires R {}                      // today: reads or write at any address; only local to module
+fun f() reads M1::R writes M2::T {}        // new access kinds -- note reference to types from other modules
+fun f<T> acquires R<T> {}                  // type instantiations supported
+fun f() acquires 0x42::*                   // wildcards
+fun f(x: T) acquires *(address_of(x))      // data dependency
+fun f() acquires !0x1::* {}                // negation
+```
+
+### File Format
+
+The existing `FunctionHandle` is extended by a field which contains an optional list of access specifiers. Each access specifier consists of a kind (read/write/..), a resource specifier, and an address specifier. Both those data types are chosen to represent the various forms of patterns as seen in the source language. For details, see the implementation.
+
+### Semantics
+
+#### Access Specifier Subsumption
+
+The conceptual syntax of access specifiers is as follows:
+
+```
+AccessSpecifier := Kind ResourceSpecifier AddressSpecifier
+ResourceSpecifier := * | Address::* | Address::Module::* | Address::Module::Resource [ TypeArgs ]
+AddressSpecifier := * | Address | Parameter | Function Parameter
+Kind := acquires | reads | writes
+```
+
+The basic function specifying the semantics of access specifiers is subsumption. One access specifier subsumes another one if all access allowed in the other are also allowed in the subsuming one. For `ResoureSpecifier` this looks as below. Notice that transitivity of subsumption is assumed:
+
+```
+*         >=     _
+A::*      >=     A::M::*
+A::M::*   >=     A::M::R
+A::M::R   >=     A::M::R<TS>
+```
+
+Subsumption for `AddressSpecifier` is similar but depends on an environment to obtain values for parameters and specific functions for them. Assume this environment is called `E`:
+
+```
+*         >=     _
+x         >=     y       iff E(x) == E(y)
+f(x)      >=     g(y)    iff E(f(x)) == E(g(y))
+```
+
+Notice that the functions which are allowed in address specifiers are selected from a small set of well-known builtins, like e.g. obtaining the address of a signer or of an Aptos object. No arbitrary Move code is executed.
```

Comment:
Can we explicitly relate this to some of the Aptos Object operators here?

Actually, I see a weakness of this formulation w.r.t. Aptos Objects.  The whole object model presumes that addresses created by well-known builtins like create_object_from_account() do not alias anything else.  Given that, we should be able to assume that addresses either provided explicitly or created by other means will not alias addresses from create_object_from_account().  Can we represent this usefully?  I have a function that can won't be anything generated by create_object_from_account(*) nor will it be certain other functions.  Perhaps something like `literal(*)` == addresses disjoint from the range of any builtin function.

Reply:
I'm not sure whether I get the point. Object address can be aliased. In `fun f(o1: Object<T>, o2: Object<T>)` the both parameters can be the same or different objects. The creation is a different story, but this model is not for creation, but for read/write.

Reply:
The case that is interesting is this: can we determine that a function which reads/writes an "object" definitely does not conflict with a function that reads/writes a non-object resource?  Or do we assume that the Resource type will be different for those two cases so that can be used to disambiguate?

For example, here is some  [example code I found](https://aptos.dev/standards/aptos-object/) referring to objects:
```
public entry fun modify_reserves(liquidity_pool: Object<LiquidityPool>) {
    let liquidity_pool = &mut borrow_global_mut<LiquidityPool>(liquidity_pool);
    liquidity_pool.reserves_a = liquidity_pool.reserves_a + 1000;
}
```
What are the appropriate access specs for this function?

Reply:
Access spec for this function would be `writes LiquidityPool(object::owner(liquidity_pool))`. Note that this example has a bug. You can only call borrow on addresses. It should have been

```
    let liquidity_pool = &mut borrow_global_mut<LiquidityPool>(object::owner(liquidity_pool));
``` 

In general, I don't think you have a way to determine non-overlapping accesses, unless with a complex reasoning machine. Specifically at runtime, this is not possible. In fact when this function is called, `writes LiquidityPool(object::owner(liquidity_pool))` will be simplified based based on the concrete parameter to `writes LiquidityPool(0x42)`.